cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -639,12 +639,8 @@ class InputState : public util::SerialSequencingQueue::Processor {
         // hit the end of the batch, need to get the next batch if possible.
         ++batches_processed_;
         latest_ref_row_ = 0;
-        have_active_batch &= !queue_.TryPop();
```

Comment:
IMO this code is apparently NOT doing what it is supposed to do: by here the `have_active_batch` must be `true`, meaning `queue_` must be non-empty. Then popping the queue always gets a valid `std::optional` and effectively setting `have_active_batch` to false. The `if` branch below won't be executed.

Reply:
I noticed there are two fixes made before in the following `if` branch #36094 and #36499. Given that this `have_active_batch &= !queue_.TryPop()` exists since #13028 where it was introduced, and the lack of corresponding tests, I'm not sure how they fixed anything and what the bugs even were. So I just remove them.